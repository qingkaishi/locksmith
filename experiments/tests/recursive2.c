#include <pthread.h>
#include <stdio.h>

int *j;

pthread_mutex_t* rec() {
  int i;
  pthread_mutex_t *t;
  if(i) {
    return rec();
  } else {
    pthread_mutex_init(t,0);
    return t;
  }
}

pthread_mutex_t *l;// = rec();
void * dostuff(void * a) {
  int k;
  
  pthread_mutex_lock(l);
  j = &k;
  pthread_mutex_unlock(l);
  return NULL;
}

int main() {
  pthread_t t1;
  int i = 1;
  l = rec();
  while (1)
  pthread_create(&t1, NULL, dostuff, NULL);
  //pthread_create(&t1, NULL, dostuff, NULL);
}


