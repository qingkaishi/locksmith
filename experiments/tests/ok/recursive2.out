************************* STARTING *************************

shared: &j:recursive2.c:4
&j:recursive2.c:4 is protected by:
  *l:recursive2.c:17
  *rec.return:recursive2.c:30
  concrete *t::-1 -> rec recursive2.c:30
  concrete *t::-1 -> rec recursive2.c:30 -> rec recursive2.c:10
*************************** DONE ***************************
